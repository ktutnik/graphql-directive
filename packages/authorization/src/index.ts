import { MapperKind, mapSchema, SchemaMapper } from "@graphql-tools/utils"
import { getDirectiveValues, GraphQLInputObjectType, GraphQLInputObjectTypeConfig, GraphQLResolveInfo, GraphQLSchema, getNamedType, GraphQLInputFieldConfig, GraphQLArgumentConfig, GraphQLFieldConfigArgumentMap } from "graphql"

// ====================================================== //
// ======================== TYPES ======================= //
// ====================================================== //


interface InvocationContext {
    path: string
    parentValue: any
    resolverArgs: [any, any, any, GraphQLResolveInfo]
}

type Invocable = (value: any, ctx: InvocationContext) => Promise<any>

interface Invocation {
    invoke(value: any, ctx: InvocationContext): Promise<any>
}

class FieldInvocation implements Invocation {
    invoke(value: any, ctx: InvocationContext): Promise<any> {
        throw new Error("Method not implemented.")
    }
}

class TypeInvocation implements Invocation {
    constructor(private name: string) { }

    fields: Record<string, Invocation> = {}

    invoke(value: any, { path, parentValue, ...ctx }: InvocationContext): Promise<unknown> {
        const context = { path: path + "." + this.name, parentValue: value, ...ctx }
        return Promise.all(Object.keys(this.fields).map(key => this.fields[key].invoke(value[key], context)))
    }
}



export { GraphQLSchema }


// ====================================================== //
// ====================== FUNCTIONS ===================== //
// ====================================================== //

type MapperHook<T> = (config: GraphQLInputFieldConfig | GraphQLArgumentConfig, name: string, type: string, sch:GraphQLSchema) => T
type Invoker<T> = (args:GraphQLFieldConfigArgumentMap) => T

function createFieldArgumentInvoker<T>(hook:MapperHook<T>): [SchemaMapper, Invoker<T>] {
    return [{}, ()]
}

const transform = (schema: GraphQLSchema): GraphQLSchema => {
    const [mapper, invoke] = createFieldArgumentInvoker((config) => {

    })

    return mapSchema(schema, {
        ...mapper,
        [MapperKind.OBJECT_FIELD]: (config, fieldName, type, sch) => {
            if (fieldName === "addUser") {
                const result = await invoke(config.args!)
                return config
            }
        },

    })
}

export const createTransformer = () =>
    (schema: GraphQLSchema): GraphQLSchema => {
        return transform(schema)
    }
