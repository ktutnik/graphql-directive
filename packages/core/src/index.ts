export * from "./directive-invoker"
export * from "./validator"