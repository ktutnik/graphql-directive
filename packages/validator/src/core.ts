
export interface TypeConfig {
    kind: "Type"
    name: String
    fields: FieldConfig[]
}

export interface FieldConfig {
    kind: "Field",
    name: string
    validator: Validator
}

export enum ValidationMethod {
    AFTER = "AFTER",
    ALPHA = "ALPHA",
    ALPHANUMERIC = "ALPHANUMERIC",
    ARRAY = "ARRAY",
    ASCII = "ASCII",
    BASE64 = "BASE64",
    BEFORE = "BEFORE",
    BOOLEAN = "BOOLEAN",
    BYTE_LENGTH = "BYTE_LENGTH",
    CREDIT_CARD = "CREDIT_CARD",
    CURRENCY = "CURRENCY",
    DATA_URI = "DATA_URI",
    DECIMAL = "DECIMAL",
    DIVISIBLE_BY = "DIVISIBLE_BY",
    EMAIL = "EMAIL",
    EMPTY = "EMPTY",
    ETHEREUM_ADDRESS = "ETHEREUM_ADDRESS",
    FQDN = "FQDN",
    FLOAT = "FLOAT",
    FULL_WIDTH = "FULL_WIDTH",
    HALF_WIDTH = "HALF_WIDTH",
    HEX_COLOR = "HEX_COLOR",
    HEX_DECIMAL = "HEX_DECIMAL",
    IP = "IP",
    IP_RANGE = "IP_RANGE",
    ISBN = "ISBN",
    ISIN = "ISIN",
    ISO8601 = "ISO8601",
    ISO31661_ALPHA2 = "ISO31661_ALPHA2",
    ISO31661_ALPHA3 = "ISO31661_ALPHA3",
    ISRC = "ISRC",
    ISSN = "ISSN",
    JSON = "JSON",
    JWT = "JWT",
    LAT_LONG = "LAT_LONG",
    LENGTH = "LENGTH",
    LOWERCASE = "LOWERCASE",
    MAC_ADDRESS = "MAC_ADDRESS",
    MIME_TYPE = "MIME_TYPE",
    MONGO_ID = "MONGO_ID",
    MULTIBYTE = "MULTIBYTE",
    NUMBER = "NUMBER",
    PORT = "PORT",
    POSTAL_CODE = "POSTAL_CODE",
    REGEX = "REGEX",
    SURROGATE_PAIR = "SURROGATE_PAIR",
    UPPERCASE = "UPPERCASE",
    URL = "URL",
    UUID = "UUID",
    VARIABLE_WIDTH = "VARIABLE_WIDTH",
    WHITELISTED = "WHITELISTED",
}

export interface DirectiveArgs {
    method: ValidationMethod,
    min: number,
    max: number
}

export interface ErrorMessage {
    path: string,
    message: string
}

export type NativeValidator = (val: any) => string | true

export type Validator = (val: any) => ErrorMessage[] | true