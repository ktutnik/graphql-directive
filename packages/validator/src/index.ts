import { getDirective, MapperKind, mapSchema } from "@graphql-tools/utils"
import { defaultFieldResolver, GraphQLError, GraphQLFieldConfig, GraphQLFieldConfigArgumentMap, GraphQLFieldResolver, GraphQLInputFieldConfig, GraphQLInputObjectType, GraphQLInputType, GraphQLSchema, Kind, ListTypeNode, NamedTypeNode, NonNullTypeNode, TypeNode } from "graphql"

import { DirectiveArgs, ErrorMessage, FieldConfig, ObjectValidator, TypeConfig, Validator } from "./core"
import nativeValidators from "./validator"

const configurations: { [key: string]: TypeConfig } = {}

const typeDefs = /* GraphQL */ `
    enum ValidationMethod {
        EMAIL, ARRAY
    }
    directive @validate(
        method: ValidationMethod!, 
        min:Int, 
        max:Int
    ) repeatable on INPUT_FIELD_DEFINITION | ARGUMENT_DEFINITION
`

const transform = (schema: GraphQLSchema): GraphQLSchema => {
    const getValidateDirectives = (config: GraphQLInputFieldConfig | GraphQLFieldConfig<any, any>) => getDirective(schema, config, "validate") as DirectiveArgs[] ?? []


    const createFieldValidator = (path: string, directives: DirectiveArgs[]): ObjectValidator => {
        const fieldValidator = directives.map<Validator>(({ method, ...config }) => nativeValidators[method](config))
        const validator: ObjectValidator = (val) => {
            if (val === null || val === undefined) return true
            // validate field
            const fieldError = fieldValidator.map(v => v(val))
                .filter((x): x is string[] => x !== true)
                .flat()
            return fieldError.length > 0 ? [{ path, message: fieldError }] : true
        }
        return validator
    }

    const createObjectValidator = (path: string, fields: FieldConfig[]): ObjectValidator => {
        const validator: ObjectValidator = (val) => {
            if (val === null || val === undefined) return true
            const messages = fields.map(v => v.validator(val[v.name]))
                .filter((x): x is ErrorMessage[] => x !== true)
                .flat()
                .map<ErrorMessage>(e => ({
                    message: e.message,
                    path: [path, e.path].filter(x => x !== "").join(".")
                }))
            return messages.length > 0 ? messages : true
        }
        return validator
    }

    const composeValidator = (...validators: ObjectValidator[]): ObjectValidator => {
        return (val: any) => {
            const messages = validators.map(v => v(val))
                .filter((x): x is ErrorMessage[] => x !== true)
                .flat();
            return messages.length > 0 ? messages : true
        }
    }


    return mapSchema(schema, {
        [MapperKind.INPUT_OBJECT_FIELD]: (config, name, type, sch) => {
            const getParentConfig = (type: string) => configurations[type] ?? (configurations[type] = { fields: [], kind: "Type", name: type })
            const directives = getValidateDirectives(config)
            const dataType = config.type.toString().replace("!", "")
            const dataTypeConfig = configurations[dataType]
            if (directives.length > 0 || !!dataTypeConfig) {
                const parent = getParentConfig(type)
                const fields = (dataTypeConfig && dataTypeConfig.fields) ?? []
                const validator = composeValidator(
                    createFieldValidator(name, directives),
                    createObjectValidator(name, fields))
                parent.fields = [...parent.fields, { kind: "Field", name, validator }]
                return config
            }
        },
        [MapperKind.OBJECT_FIELD]: (config, fieldName, type, sch) => {
            const validators: FieldConfig[] = []
            for (const key in config.args) {
                const argConfig = config.args[key]
                const directives = getValidateDirectives(argConfig)
                const dataType = argConfig.type.toString().replace("!", "")
                const name = argConfig.astNode!.name.value
                const dataTypeConfig = configurations[dataType]
                if (directives.length > 0 || !!dataTypeConfig) {
                    const fields = (dataTypeConfig && dataTypeConfig.fields) ?? []
                    const validator = composeValidator(
                        createFieldValidator(name, directives),
                        createObjectValidator(name, fields))
                    validators.push({ kind: "Field", name, validator })
                }
            }
            if (validators.length > 0) {
                const validator = createObjectValidator("", validators)
                const { resolve = defaultFieldResolver } = config
                return {
                    ...config,
                    resolve: (source, args, context, info) => {
                        const error = validator(args)
                        if (error !== true) throw new GraphQLError("USER_INPUT_ERROR", { extensions: { error } })
                        return resolve(source, args, context, info)
                    }
                }
            }
        }
    })
}

export default {
    typeDefs, transform
}
