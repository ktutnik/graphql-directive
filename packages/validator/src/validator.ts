import val from "validator"
import { ValidationMethod, NativeValidator } from "./core"

type Validators = {
    [key: string]:
    (config: any) => NativeValidator

}

export default {
    [ValidationMethod.AFTER]:
        (options: { date: string }) => (str) => val.isAfter(str, options.date)
            || `Must be a date after ${options.date}`,

    [ValidationMethod.ALPHA]:
        (options: { locale: val.AlphaLocale }) => (str) => val.isAlpha(str, options.locale)
            || 'Must only contain letters',

    [ValidationMethod.ALPHANUMERIC]:
        (locale) => (str) => val.isAlphanumeric(str, locale)
            || 'Must only contain letters and numbers',

    [ValidationMethod.ASCII]:
        () => (str) => val.isAscii(str)
            || 'Must only contain ASCII characters',

    [ValidationMethod.BASE64]:
        () => (str) => val.isBase64(str)
            || 'Must be a valid base64 encoded string',

    [ValidationMethod.BEFORE]:
        (options: { date: string }) => (str) => val.isBefore(str, options.date)
            || `Must be a date before ${options.date}`,

    [ValidationMethod.BOOLEAN]:
        () => (str) => val.isBoolean(str)
            || 'Must be a boolean',

    [ValidationMethod.BYTE_LENGTH]:
        (options: { min?: number, max?: number }) => (str) => val.isByteLength(str, options)
            || 'Must have a length (in bytes) within the specified range',

    [ValidationMethod.CREDIT_CARD]:
        () => (str) => val.isCreditCard(str)
            || 'Must be a valid credit card number',

    [ValidationMethod.CURRENCY]:
        (options) => (str) => val.isCurrency(str, options)
            || 'Must be a valid currency amount',

    [ValidationMethod.DATA_URI]:
        () => (str) => val.isDataURI(str)
            || 'Must be a valid data URI',

    [ValidationMethod.DECIMAL]:
        (options) => (str) => val.isDecimal(str, options)
            || 'Must be a decimal number',

    [ValidationMethod.DIVISIBLE_BY]:
        (number) => (str) => val.isDivisibleBy(str, number)
            || `Must be a number that's divisible by ${number}`,

    [ValidationMethod.EMAIL]:
        (options) => (str) => val.isEmail(str, options)
            || 'Must be a valid email address',

    [ValidationMethod.EMPTY]:
        (options) => (str) => val.isEmpty(str, options)
            || 'Must not be empty',

    [ValidationMethod.ETHEREUM_ADDRESS]:
        () => (str) => val.isEthereumAddress(str)
            || 'Must be a valid Ethereum address',

    [ValidationMethod.FQDN]:
        (options) => (str) => val.isFQDN(str, options)
            || 'Must be a valid fully qualified domain name',

    [ValidationMethod.FLOAT]:
        (options) => (str) => val.isFloat(str, options)
            || 'Must be a float number',

    [ValidationMethod.FULL_WIDTH]:
        () => (str) => val.isFullWidth(str)
            || 'Must contain full-width characters',

    [ValidationMethod.HALF_WIDTH]:
        () => (str) => val.isHalfWidth(str)
            || 'Must contain half-width characters',

    [ValidationMethod.HEX_COLOR]:
        () => (str) => val.isHexColor(str)
            || 'Must be a valid hex color code',

    [ValidationMethod.HEX_DECIMAL]:
        () => (str) => val.isHexadecimal(str)
            || 'Must be a valid hexadecimal number',

    [ValidationMethod.IP]:
        (version) => (str) => val.isIP(str, version)
            || `Must be a valid ${version} IP address`,

    [ValidationMethod.IP_RANGE]:
        () => (str) => val.isIPRange(str)
            || 'Must be a valid IP range',

    [ValidationMethod.ISBN]:
        (version) => (str) => val.isISBN(str, version)
            || `Must be a valid ISBN ${version}`,

    [ValidationMethod.ISIN]:
        () => (str) => val.isISIN(str)
            || 'Must be a valid ISIN (International Securities Identification Number)',

    [ValidationMethod.ISO8601]:
        (options) => (str) => val.isISO8601(str, options)
            || 'Must be a valid ISO8601 date string',

    [ValidationMethod.ISO31661_ALPHA2]:
        () => (str) => val.isISO31661Alpha2(str)
            || 'Must be a valid ISO 3166-1 alpha-2 country code',

    [ValidationMethod.ISO31661_ALPHA3]:
        () => (str) => val.isISO31661Alpha3(str)
            || 'Must be a valid ISO 3166-1 alpha-3 country code',

    [ValidationMethod.ISRC]:
        () => (str) => val.isISRC(str)
            || 'Must be a valid International Standard Recording Code (ISRC)',

    [ValidationMethod.ISSN]:
        (options) => (str) => val.isISSN(str, options)
            || 'Must be a valid International Standard Serial Number (ISSN)',

    [ValidationMethod.JSON]:
        () => (str) => val.isJSON(str)
            || 'Must be a valid JSON string',

    [ValidationMethod.JWT]:
        () => (str) => val.isJWT(str)
            || 'Must be a valid JSON Web Token (JWT)',

    [ValidationMethod.LAT_LONG]:
        () => (str) => val.isLatLong(str)
            || 'Must be a valid latitude coordinate',

    [ValidationMethod.LENGTH]:
        (options) => (str) => val.isLength(str, options)
            || 'Must have a length within the specified range',

    [ValidationMethod.LOWERCASE]:
        () => (str) => val.isLowercase(str)
            || 'Must be all lowercase',

    [ValidationMethod.MAC_ADDRESS]:
        () => (str) => val.isMACAddress(str)
            || 'Must be a valid MAC address',

    [ValidationMethod.MIME_TYPE]:
        () => (str) => val.isMimeType(str)
            || 'Must be a valid MIME type',

    [ValidationMethod.MONGO_ID]:
        () => (str) => val.isMongoId(str)
            || 'Must be a valid MongoDB ObjectId',

    [ValidationMethod.MULTIBYTE]:
        () => (str) => val.isMultibyte(str)
            || 'Must contain one or more multibyte characters',

    [ValidationMethod.NUMBER]:
        (options) => (str) => val.isNumeric(str, options)
            || 'Must be a number',

    [ValidationMethod.PORT]:
        () => (str) => val.isPort(str)
            || 'Must be a valid port number',

    [ValidationMethod.POSTAL_CODE]:
        (locale) => (str) => val.isPostalCode(str, locale)
            || 'Must be a valid postal code',

    [ValidationMethod.URL]:
        (options) => (str) => val.isURL(str, options)
            || 'Must be a valid URL',

    [ValidationMethod.UUID]:
        (version) => (str) => val.isUUID(str, version)
            || 'Must be a valid UUID',

    [ValidationMethod.VARIABLE_WIDTH]:
        () => (str) => val.isVariableWidth(str)
            || 'Must contain a mixture of full and half-width characters',

    [ValidationMethod.WHITELISTED]:
        (chars) => (str) => val.isWhitelisted(str, chars)
            || `Must only contain characters from the whitelist: ${chars}`,

    [ValidationMethod.SURROGATE_PAIR]:
        () => (str) => val.isSurrogatePair(str)
            || 'Must contain any surrogate pairs characters',

    [ValidationMethod.UPPERCASE]:
        () => (str) => val.isUppercase(str)
            || 'Must only contain uppercase characters',

    [ValidationMethod.ARRAY]:
        ({ min, max }: { min?: number, max?: number }) => (arr: any[]) => {

            if (!arr) return true
            if (min && arr.length < min) return `Must have at least ${min} items`
            if (max && arr.length > max) return `Must have less than ${max} items`
            return true
        }
} as Validators
