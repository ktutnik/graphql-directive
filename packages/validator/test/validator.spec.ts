import { ValidationMethod } from "../src/core"

import val from "../src/validator"


describe("Validator", () => {


    it("Should validate date properly", () => {
        const validate = val[ValidationMethod.AFTER]({ date: "2003-1-1" })

        expect(validate("2004-1-1")).toBe(true);
        expect(validate("2001-1-1")).toMatchSnapshot();
    });

    it("Should validate alphabetic string properly", () => {
        const validate = val[ValidationMethod.ALPHA]({})

        expect(validate("abc")).toBe(true);
        expect(validate("123")).toMatchSnapshot()
    });

    it("Should validate alphanumeric string properly", () => {
        const validate = val[ValidationMethod.ALPHANUMERIC]({})

        expect(validate("abc123")).toBe(true);
        expect(validate("abc!@#")).toMatchSnapshot()
    });

    it("Should validate ASCII string properly", () => {
        const validate = val[ValidationMethod.ASCII]({})

        expect(validate("Hello world")).toBe(true);
        expect(validate("你好世界")).toMatchSnapshot()
    });

    it("Should validate base64 string properly", () => {
        const validate = val[ValidationMethod.BASE64]({})

        expect(validate("SGVsbG8gV29ybGQh")).toBe(true);
        expect(validate("Not a base64 string!")).toMatchSnapshot()
    });

    it("Should validate date before properly", () => {
        const validate = val[ValidationMethod.BEFORE]({ date: "2023-01-01" })

        expect(validate("2022-12-31")).toBe(true);
        expect(validate("2023-01-01")).toMatchSnapshot()
    });

    it("Should validate boolean properly", () => {
        const validate = val[ValidationMethod.BOOLEAN]({})

        expect(validate("true")).toBe(true);
        expect(validate("false")).toBe(true);
        expect(validate("notaboolean")).toMatchSnapshot()
    });

    it("Should validate credit card properly", () => {
        const validate = val[ValidationMethod.CREDIT_CARD]({})

        expect(validate("4111111111111111")).toBe(true);
        expect(validate("notacreditcardnumber")).toMatchSnapshot()
    });

    it("Should validate currency properly", () => {
        const validate = val[ValidationMethod.CURRENCY]({})

        expect(validate("$1.00")).toBe(true);
        expect(validate("notacurrency")).toMatchSnapshot()
    });

    it("Should validate data URI properly", () => {
        const validate = val[ValidationMethod.DATA_URI]({})

        expect(validate("data:image/png;base64,iVBORw")).toBe(true);
        expect(validate("notadataURI")).toMatchSnapshot()
    });

    it("Should validate decimal properly", () => {
        const validate = val[ValidationMethod.DECIMAL]({})

        expect(validate("1234.5678")).toBe(true);
        expect(validate("1,234")).toMatchSnapshot()
    });

    it("Should validate divisible properly", () => {
        const validate = val[ValidationMethod.DIVISIBLE_BY]({ number: 3 })

        expect(validate("9")).toBe(true);
        expect(validate("10")).toMatchSnapshot()
    });

    it("Should validate email properly", () => {
        const validate = val[ValidationMethod.EMAIL]({})

        expect(validate("test@example.com")).toBe(true);
        expect(validate("notanemail")).toMatchSnapshot()
    });

    it("Should validate empty properly", () => {
        const validate = val[ValidationMethod.EMPTY]({})

        expect(validate("")).toBe(true);
        expect(validate("hello")).toMatchSnapshot()
    });

    it("Should validate Ethereum address properly", () => {
        const validate = val[ValidationMethod.ETHEREUM_ADDRESS]({})

        expect(validate("0x742d35Cc6634C0532925a3b844Bc454e4438f44e")).toBe(true);
        expect(validate("notanethereumaddress")).toMatchSnapshot()
    });


    it("Should validate domain name properly", () => {
        const validate = val[ValidationMethod.FQDN]({})

        expect(validate("example.com")).toBe(true);
        expect(validate("example")).toMatchSnapshot()
    });

    it("Should validate float properly", () => {
        const validate = val[ValidationMethod.FLOAT]({})

        expect(validate("1.23")).toBe(true);
        expect(validate("invalidnumber")).toMatchSnapshot()
    });

    it("Should validate full-width string properly", () => {
        const validate = val[ValidationMethod.FULL_WIDTH]({})

        expect(validate("ＨＥＬＬＯ")).toBe(true);
        expect(validate("hello")).toMatchSnapshot()
    });

    it("Should validate half-width string properly", () => {
        const validate = val[ValidationMethod.HALF_WIDTH]({})

        expect(validate("hello")).toBe(true);
        expect(validate("ＨＥＬＬＯ")).toMatchSnapshot()
    });

    it("Should validate hexadecimal color code properly", () => {
        const validate = val[ValidationMethod.HEX_COLOR]({})

        expect(validate("#ff0000")).toBe(true);
        expect(validate("notacolorcode")).toMatchSnapshot()
    });

    it("Should validate hexadecimal properly", () => {
        const validate = val[ValidationMethod.HEXADECIMAL]({})

        expect(validate("deadbeef")).toBe(true);
        expect(validate("notahexadecimal")).toMatchSnapshot()
    });

    it("Should validate IP address properly", () => {
        const validate = val[ValidationMethod.IP]({ version: 4 })

        expect(validate("192.168.1.1")).toBe(true);
        expect(validate("notanipaddress")).toMatchSnapshot()
    });


    it("Should validate IP range properly", () => {
        const validate = val[ValidationMethod.IP_RANGE]({})

        expect(validate("192.168.0.1/24")).toBe(true);
        expect(validate("notaniprange")).toMatchSnapshot()
    });

    it("Should validate ISBN properly", () => {
        const validate = val[ValidationMethod.ISBN]({ version: 10 })

        expect(validate("3-8362-2119-5")).toBe(true);
        expect(validate("notanisbn")).toMatchSnapshot()
    });

    it("Should validate ISIN properly", () => {
        const validate = val[ValidationMethod.ISIN]({})

        expect(validate("US0378331005")).toBe(true);
        expect(validate("notanisin")).toMatchSnapshot()
    });

    it("Should validate ISO8601 date properly", () => {
        const validate = val[ValidationMethod.ISO8601]({})

        expect(validate("2022-03-02")).toBe(true);
        expect(validate("notaniso8601date")).toMatchSnapshot()
    });

    it("Should validate ISO 3166-1 alpha-2 code properly", () => {
        const validate = val[ValidationMethod.ISO31661_ALPHA2]({})

        expect(validate("US")).toBe(true);
        expect(validate("notaniso31661alpha2code")).toMatchSnapshot()
    });

    it("Should validate ISO 3166-1 alpha-3 code properly", () => {
        const validate = val[ValidationMethod.ISO31661_ALPHA3]({})

        expect(validate("USA")).toBe(true);
        expect(validate("notaniso31661alpha3code")).toMatchSnapshot()
    });

    it("Should validate ISRC properly", () => {
        const validate = val[ValidationMethod.ISRC]({})

        expect(validate("USRC17607839")).toBe(true);
        expect(validate("notanisrc")).toMatchSnapshot()
    });

    it("Should validate ISSN properly", () => {
        const validate = val[ValidationMethod.ISSN]({})

        expect(validate("0317-8471")).toBe(true);
        expect(validate("notanissn")).toMatchSnapshot()
    });

    it("Should validate JSON properly", () => {
        const validate = val[ValidationMethod.JSON]({})

        expect(validate('{"name":"John","age":30}')).toBe(true);
        expect(validate("notajsonstring")).toMatchSnapshot()
    });


    it("Should validate JWT properly", () => {
        const validate = val[ValidationMethod.JWT]({})

        expect(
            validate(
                "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiIxMjM0NTY3ODkwIiwibmFtZSI6IkpvaG4gRG9lIiwiaWF0IjoxNTE2MjM5MDIyfQ.SflKxwRJSMeKKF2QT4fwpMeJf36POk6yJV_adQssw5c"
            )
        ).toBe(true);
        expect(validate("notajwt")).toMatchSnapshot()
    });

    it("Should validate latitude and longitude properly", () => {
        const validate = val[ValidationMethod.LAT_LONG]({})

        expect(validate("55.720923,-28.652344")).toBe(true);
        expect(validate("90.1000000, 180.000000")).toMatchSnapshot()
    });

    it("Should validate length properly", () => {
        const validate = val[ValidationMethod.LENGTH]({ min: 3, max: 5 })

        expect(validate("1234")).toBe(true);
        expect(validate("123456")).toMatchSnapshot()
    });

    it("Should validate lowercase properly", () => {
        const validate = val[ValidationMethod.LOWERCASE]({})

        expect(validate("hello")).toBe(true);
        expect(validate("Hello")).toMatchSnapshot()
    });

    it("Should validate MAC address properly", () => {
        const validate = val[ValidationMethod.MAC_ADDRESS]({})

        expect(validate("01:23:45:67:89:ab")).toBe(true);
        expect(validate("notamacaddress")).toMatchSnapshot()
    });

    it("Should validate MIME type properly", () => {
        const validate = val[ValidationMethod.MIME_TYPE]({})

        expect(validate("image/jpeg")).toBe(true);
        expect(validate("notamimetype")).toMatchSnapshot()
    });

    it("Should validate MongoDB ObjectId properly", () => {
        const validate = val[ValidationMethod.MONGO_ID]({})

        expect(validate("507f1f77bcf86cd799439011")).toBe(true);
        expect(validate("notamongodbobjectid")).toMatchSnapshot()
    });

    it("Should validate multibyte characters properly", () => {
        const validate = val[ValidationMethod.MULTIBYTE]({})

        expect(validate("こんにちは")).toBe(true);
        expect(validate("hello")).toMatchSnapshot()
    });

    it("Should validate numeric characters properly", () => {
        const validate = val[ValidationMethod.NUMBER]({})

        expect(validate("1234")).toBe(true);
        expect(validate("notanumber")).toMatchSnapshot()
    });

    it("Should validate port number properly", () => {
        const validate = val[ValidationMethod.PORT]({})

        expect(validate("3000")).toBe(true);
        expect(validate("notaport")).toMatchSnapshot()
    });

    it("Should validate postal code properly", () => {
        const validate = val[ValidationMethod.POSTAL_CODE]({ locale: "US" })

        expect(validate("94043")).toBe(true);
        expect(validate("notapostalcode")).toMatchSnapshot()
    });

    it("Should validate URL properly", () => {
        const validate = val[ValidationMethod.URL]({})

        expect(validate("https://example.com")).toBe(true);
        expect(validate("notanurl")).toMatchSnapshot()
    });

    it("Should validate UUID properly", () => {
        const validate = val[ValidationMethod.UUID]({ version: "all" })

        expect(validate("f7eb504d-b7e3-4e3f-a87d-1d93b98df94d")).toBe(true);
        expect(validate("notauuid")).toMatchSnapshot()
    });


    it('should validate variable width characters', () => {
        const validate = val[ValidationMethod.VARIABLE_WIDTH]({})

        expect(validate('ひらがなカタカナ漢字ABCDE')).toBe(true);
        expect(validate('12345')).toMatchSnapshot()
        expect(validate('abcde')).toMatchSnapshot()
    });

    it('should validate whitelisted characters', () => {
        const validate = val[ValidationMethod.WHITELISTED]({ chars: 'abcdefghijklmnopqrstuvwxyz' })

        expect(validate('abcdefghijklmnopqrstuvwxyz')).toBe(true);
        expect(validate('ABCDEFGHIJKLMNOPQRSTUVWXYZ')).toMatchSnapshot()
        expect(validate('0123456789')).toMatchSnapshot()
        expect(validate('aBc123')).toMatchSnapshot()
        expect(validate('')).toBe(true);
    });

    it('should validate surrogate pairs', () => {
        const validate = val[ValidationMethod.SURROGATE_PAIR]({})
        expect(validate('\uD83D\uDC68\uD83C\uDFFB')).toBe(true);
        expect(validate('')).toMatchSnapshot()
    });

    it('should validate uppercase characters', () => {
        const validate = val[ValidationMethod.UPPERCASE]({})
        expect(validate('ABCDEFGHIJKLMNOPQRSTUVWXYZ')).toBe(true);
        expect(validate('abcdefghijklmnopqrstuvwxyz')).toMatchSnapshot()
        expect(validate('aBcDeF')).toMatchSnapshot()
    });


})


