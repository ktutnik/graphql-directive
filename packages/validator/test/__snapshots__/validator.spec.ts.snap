// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Validator Should validate ASCII string properly 1`] = `"Must only contain ASCII characters"`;

exports[`Validator Should validate Ethereum address properly 1`] = `"Must be a valid Ethereum address"`;

exports[`Validator Should validate IP address properly 1`] = `"Must be a valid 4 IP address"`;

exports[`Validator Should validate IP range properly 1`] = `"Must be a valid IP range"`;

exports[`Validator Should validate ISBN properly 1`] = `"Must be a valid ISBN 10"`;

exports[`Validator Should validate ISIN properly 1`] = `"Must be a valid ISIN (International Securities Identification Number)"`;

exports[`Validator Should validate ISO 3166-1 alpha-2 code properly 1`] = `"Must be a valid ISO 3166-1 alpha-2 country code"`;

exports[`Validator Should validate ISO 3166-1 alpha-3 code properly 1`] = `"Must be a valid ISO 3166-1 alpha-3 country code"`;

exports[`Validator Should validate ISO8601 date properly 1`] = `"Must be a valid ISO8601 date string"`;

exports[`Validator Should validate ISRC properly 1`] = `"Must be a valid International Standard Recording Code (ISRC)"`;

exports[`Validator Should validate ISSN properly 1`] = `"Must be a valid International Standard Serial Number (ISSN)"`;

exports[`Validator Should validate JSON properly 1`] = `"Must be a valid JSON string"`;

exports[`Validator Should validate JWT properly 1`] = `"Must be a valid JSON Web Token (JWT)"`;

exports[`Validator Should validate MAC address properly 1`] = `"Must be a valid MAC address"`;

exports[`Validator Should validate MIME type properly 1`] = `"Must be a valid MIME type"`;

exports[`Validator Should validate MongoDB ObjectId properly 1`] = `"Must be a valid MongoDB ObjectId"`;

exports[`Validator Should validate URL properly 1`] = `"Must be a valid URL"`;

exports[`Validator Should validate UUID properly 1`] = `"Must be a valid UUID"`;

exports[`Validator Should validate alphabetic string properly 1`] = `"Must only contain letters"`;

exports[`Validator Should validate alphanumeric string properly 1`] = `"Must only contain letters and numbers"`;

exports[`Validator Should validate base64 string properly 1`] = `"Must be a valid base64 encoded string"`;

exports[`Validator Should validate boolean properly 1`] = `"Must be a boolean"`;

exports[`Validator Should validate credit card properly 1`] = `"Must be a valid credit card number"`;

exports[`Validator Should validate currency properly 1`] = `"Must be a valid currency amount"`;

exports[`Validator Should validate data URI properly 1`] = `"Must be a valid data URI"`;

exports[`Validator Should validate date before properly 1`] = `"Must be a date before 1/1/2023"`;

exports[`Validator Should validate date properly 1`] = `"Must be a date after 1/1/2003"`;

exports[`Validator Should validate decimal properly 1`] = `"Must be a decimal number"`;

exports[`Validator Should validate divisible properly 1`] = `"Must be a number that's divisible by 3"`;

exports[`Validator Should validate domain name properly 1`] = `"Must be a valid fully qualified domain name"`;

exports[`Validator Should validate email properly 1`] = `"Must be a valid email address"`;

exports[`Validator Should validate empty properly 1`] = `"Must not be empty"`;

exports[`Validator Should validate float properly 1`] = `"Must be a float number"`;

exports[`Validator Should validate full-width string properly 1`] = `"Must contain full-width characters"`;

exports[`Validator Should validate half-width string properly 1`] = `"Must contain half-width characters"`;

exports[`Validator Should validate hexadecimal color code properly 1`] = `"Must be a valid hex color code"`;

exports[`Validator Should validate hexadecimal properly 1`] = `"Must be a valid hexadecimal number"`;

exports[`Validator Should validate latitude and longitude properly 1`] = `"Must be a valid latitude coordinate"`;

exports[`Validator Should validate length properly 1`] = `"Must have a length within the specified range"`;

exports[`Validator Should validate lowercase properly 1`] = `"Must be all lowercase"`;

exports[`Validator Should validate multibyte characters properly 1`] = `"Must contain one or more multibyte characters"`;

exports[`Validator Should validate numeric characters properly 1`] = `"Must be a number"`;

exports[`Validator Should validate port number properly 1`] = `"Must be a valid port number"`;

exports[`Validator Should validate postal code properly 1`] = `"Must be a valid postal code"`;

exports[`Validator should validate surrogate pairs 1`] = `"Must contain any surrogate pairs characters"`;

exports[`Validator should validate uppercase characters 1`] = `"Must only contain uppercase characters"`;

exports[`Validator should validate uppercase characters 2`] = `"Must only contain uppercase characters"`;

exports[`Validator should validate variable width characters 1`] = `"Must contain a mixture of full and half-width characters"`;

exports[`Validator should validate variable width characters 2`] = `"Must contain a mixture of full and half-width characters"`;

exports[`Validator should validate whitelisted characters 1`] = `"Must only contain characters from the whitelist: abcdefghijklmnopqrstuvwxyz"`;

exports[`Validator should validate whitelisted characters 2`] = `"Must only contain characters from the whitelist: abcdefghijklmnopqrstuvwxyz"`;

exports[`Validator should validate whitelisted characters 3`] = `"Must only contain characters from the whitelist: abcdefghijklmnopqrstuvwxyz"`;
